src/libcretonne/cfg.rs

```diff
@@ -0,0 +1,270 @@
+    data: BTreeMap<Ebb, PredecessorSet>,
+}
+
+impl ControlFlowGraph {
+
+    /// During initialization mappings will be generated for any existing
+    /// blocks within the CFG's associated function. Basic sanity checks will
+    /// also be performed to ensure that the blocks are well formed.
+    pub fn new(func: &Function) -> Result<ControlFlowGraph, String> {
+        let mut cfg = ControlFlowGraph{data: BTreeMap::new()};
+
+        // Even ebbs without predecessors should show up in the CFG, albeit
+        // with no entires.
+        for ebb in func.ebbs_numerically() {
+            try!(cfg.init_ebb(ebb));
+        }
+
+        for ebb in func.ebbs_numerically() {
+            // Flips to true when a terminating instruction is seen. So that if additional
+            // instructions occur an error may be returned.
+            let mut terminated = false;
+            for inst in func.ebb_insts(ebb) {
+                if terminated {
+                    return Err(format!("{} contains unreachable instructions.", ebb));
+                }
+
+                match func[inst] {
+                    InstructionData::Branch { ty: _, opcode: _, ref data } => {
+                        try!(cfg.add_predecessor(data.destination, (ebb, inst)));
+                    }
+                    InstructionData::Jump { ty: _, opcode: _, ref data } => {
+                        try!(cfg.add_predecessor(data.destination, (ebb, inst)));
+                        terminated = true;
+                    }
+                    InstructionData::Return { ty: _, opcode: _, data: _ } => {
+                        terminated = true;
+                    }
+                    InstructionData::Nullary { ty: _, opcode: _ } => {
+                        terminated = true;
+                    }
+                    _ => ()
+                }
+            }
+        }
+        Ok(cfg)
+    }
+
+    /// Initializes a predecessor set for some ebb. If an ebb already has an
+    /// entry it will be clobbered.
+    pub fn init_ebb(&mut self, ebb: Ebb) -> Result<&mut PredecessorSet, &'static str> {
+        self.data.insert(ebb, BTreeSet::new());
+        match self.data.get_mut(&ebb) {
+            Some(predecessors) => Ok(predecessors),
+            None => Err("Ebb initialization failed.")
+        }
+    }
+
+    /// Attempts to add a predecessor for some ebb, attempting to initialize
+    /// any ebb which has no entry.
+    pub fn add_predecessor(&mut self, ebb: Ebb, predecessor: Predecessor) -> Result<(), &'static str> {
+        let success = match self.data.get_mut(&ebb) {
+            Some(predecessors) => predecessors.insert(predecessor),
+            None => false
+        };
+
+        if success {
+            Ok(())
+        } else {
+            let mut predecessors = try!(self.init_ebb(ebb));
+            if predecessors.insert(predecessor) {
+                return Ok(());
+            }
+            Err("Predecessor insertion failed.")
```

Comment:
I think this can safely be an assertion, and then this function and `new` amd `init_ebb` don't ned to return results.